Interface update based on GETTSIM 1.0 (#7)
ttsim-dev/gettsim-personas: README.md

persona exists that corresponds to your use case, it provides a minimal set of input
data, overriding nodes of the tax-transfer system that are probably not relevant for
your use case (e.g. the calculation of pension benefits when you're interested in the
income tax).

If no existing persona corresponds to your use case, feel free to open an
[issue](https://github.com/ttsim-dev/gettsim-personas/issues) or
[make a contribution](https://gettsim.readthedocs.io/en/stable/gettsim_developer/how-to-contribute.html)!

## Basic example

We first show the simplest case of loading a persona and running GETTSIM on it.

Personas must be instantiated with a `policy_date`, as their content varies depending on
the policy environment at that date.

We start by importing the module with personas relevant to calculating net income for
working-age people who do not take up (or qualify for) any means- or health-tested
transfers:

```python
from gettsim_personas import einkommensteuer_sozialabgaben

example_persona = einkommensteuer_sozialabgaben.Couple1Child(
    policy_date_str="2025-01-01"
)
```

`example_persona` is a `Persona` object with the following attributes:

> - `description`: A description of the persona. Use this to check if the persona is

Author: No, the `description` attribute is really just a string. Do you have a different impression when reading this paragraph?

> - `description`: A description of the persona. Use this to check if the persona is

Author: Ah yes, fixed it!

- `input_data_tree`: The input data tree of this persona. This can be passed directly to
  GETTSIM's `main` function.
- `tt_targets_tree`: The targets that can be computed for this persona. This can also be
  passed to GETTSIM's `main` function.

> [!WARNING]
> Be careful when using personas outside their intended context. Many personas overwrite
> GETTSIM's policy functions. Always check whether a persona is suitable for your use
> case. Before using a persona, review its `description` field to ensure it fits your
> needs. For example, the personas in the `einkommensteuer_sozialabgaben` package are
> not suited to compute disposable income of low-income households, because the persona
> inputs are set up in a way that all means- and health-tested transfers are assumed not
> to be taken up (while in reality, some of those transfers have very high take-up rates
> among low-income households).

You can now compute taxes and transfers for the selected persona:

```python
from gettsim import main, MainTarget, InputData, TTTargets

result = main(
    main_target=MainTarget.results.df_with_nested_columns,
    policy_date=example_persona.policy_date,

>     input_data=InputData.tree(example_persona.input_data_tree),

Author: In GETTSIM, we have an `einnahmen` namespace which is supposed to collect all incomes that are considered as Einnahmen from the perspective of the EStG (Basically "Einnahmen" is gross income without any deductions, see [this](https://upload.wikimedia.org/wikipedia/commons/d/df/Einnahmen_Einkuenfte_Einkommen_zvE.png) figure).  We don't have "Einnahmen" for income from self-employment because GETTSIM will never be able to do the gains/losses calculation necessary to obtain "Einkünfte". Instead, income from self-employment is an input in the `("einkommensteuer", "einkünfte")` namespace.

>     input_data=InputData.tree(example_persona.input_data_tree),

Reviewer: That makes a lot of sense, thank you!

    tt_targets=TTTargets.tree(example_persona.tt_targets_tree),
)

print(result)

> >>>      einkommensteuer kindergeld    sozialversicherung

Reviewer 2: I guess the markdown-export does not play nicely with the hierarchical column names. So we should have a note somewhere that this should become a notebook. We could actually just add it to gettsim/docs, no problem to make the rtd environment depend on gettsim-personas.

Not for this PR, ofc.

>>>          betrag_y_sn   betrag_y          arbeitslosen               kranken                pflege                 rente
>>>                  NaN        NaN               beitrag               beitrag               beitrag               beitrag
>>>                  NaN        NaN betrag_versicherter_y betrag_versicherter_y betrag_versicherter_y betrag_versicherter_y
>>> p_id
>>> 0             7164.0       3060                 468.0                3078.0                 648.0                3348.0
>>> 1             7164.0          0                 468.0                3078.0                 648.0                3348.0
>>> 2                0.0          0                   0.0                   0.0                   0.0                   0.0
```

### Optional Arguments When Instantiating a Persona

You can provide a grid of earnings levels to compute taxes and transfers across
different earnings levels. For example, we may want to look at net income in our example

> household where the father earns 4,000 Euro per month. If working full-time, the mother

Author: Ah, I think you looked at the persona without modifications? Both partners earn 3000€ there. This paragraph is about the upserted persona.

would earn the same amount. We want to investigate how total net income varies as she

> changes her weekly ours between 25% and 75% of full-time work.

Reviewer 2: ```suggestion
changes her weekly hours between 25% and 75% of full-time work.
```

    policy_date_str="2025-01-01"
)
```

Next, define the input data to upsert:

```python
rent_to_upsert = {
    "wohnen": {"bruttokaltmiete_m_hh": [600.0, 600.0, 600.0, 800.0, 800.0, 800.0]}
}
```

Alternatively, you can generate a range of rent levels using `numpy.linspace`:

```python
import numpy as np

rent_to_upsert = {
    "wohnen": {
        "bruttokaltmiete_m_hh": [
            x for i in np.linspace(300, 1800, 601) for x in [i, i, i]
        ],
    }
}
```

> [!WARNING]
> The order of input data matters! GETTSIM uses pointers to `p_id`s in the input data to
> depict household structures and relationships between individuals. In general,
> `[0.0, 0.0, 4000.0, 0.0, 0.0, 6000.0]` will yield completely different results than
> `[4000.0, 0.0, 0.0, 6000.0, 0.0, 0.0]`. Always check the persona's household structure
> carefully before modifying input data. In the example above, inputs are on the
> household level, so every household member should have the same value for
> `bruttokaltmiete_m_hh`.

Now, we create a new persona object based on the original one, but with the modified
input data:

```python

> upserted_persona = basic_subsistence_benefit_persona.upsert_input_data(

Reviewer 2: ```suggestion
persona_with_varying_rent = basic_subsistence_benefit_persona.upsert_input_data(
```

persona_jan_2025 = Couple1Child(
    policy_date="2025-01-01",
    bruttolohn_m_linspace_grid=Couple1Child.LinspaceGrid(
        p0=Couple1Child.LinspaceRange(bottom=0, top=10000),
        p1=Couple1Child.LinspaceRange(bottom=0, top=10000),
        p2=Couple1Child.LinspaceRange(bottom=0, top=0),
        n_points=100,
    ),
)
```

### Computing Taxes and Transfers

You can now compute taxes and transfers for the selected persona:

```python
from gettsim import main, MainTarget, InputData, TTTargets

result = main(
    main_target=MainTarget.results.df_with_nested_columns,
    policy_date_str=jan_2025,
    input_data=InputData.tree(persona_jan_2025.input_data_tree),
    tt_targets=TTTargets.tree(persona_jan_2025.tt_targets_tree),
)
```

### Advanced Usage: Upserting Input Data

You can also vary persona input data across dimensions other than earnings. The
`upsert_input_data` function creates copies of the persona's input data, varying them
over the dimensions you specify, while preserving the household structure of the
original persona.

#### How to Upsert Input Data

Suppose you are interested in households that receive basic subsistence benefits for the

> unemployed (citizen's income). You want to vary their benefit entitlement by changing

Reviewer 2: ```suggestion
unemployed (Bürgergeld, formerly known as Arbeitslosengeld 2). You want to vary their benefit entitlement by changing
```

their gross rent excluding dwelling costs (a GETTSIM input variable).

First, instantiate the base persona:

```python

> from gettsim_personas.grundsicherung_für_erwerbsfähige import Couple1Child

Reviewer 2: ```suggestion
from gettsim_personas import grundsicherung_für_erwerbsfähige
```

I even made the above comment before seeing this...

basic_subsistence_benefit_persona = Couple1Child(policy_date="2025-01-01")
```

Next, define the input data to upsert:

```python
rent_to_upsert = {
    "wohnen": {"bruttokaltmiete_m_hh": [600.0, 600.0, 600.0, 800.0, 800.0, 800.0]}
}
```

Alternatively, you can generate a range of rent levels using `numpy.linspace`:

```python
import numpy as np

rent_to_upsert = {
    "wohnen": {
        "bruttokaltmiete_m_hh": [

>             x for i in np.linspace(300, 1800, 601) for x in [i, 0.0, 0.0]

Reviewer 2: Does this work as intended or is it a copy-paste leftover?

Author: Works as intended.

Reviewer 2: But what do we have the two 0.0 for?

        ],
    }
}
```

> [!WARNING]

> > The order of input data matters! GETTSIM uses pointers to `p_id`s in the input data to

Reviewer 2: Explain that this is irrelevant here because we are using a household-level input.

> depict household structures and relationships between individuals. In general,
> `[0.0, 0.0, 4000.0, 0.0, 0.0, 6000.0]` will yield completely different results than
> `[4000.0, 0.0, 0.0, 6000.0, 0.0, 0.0]`. Always check the persona's household structure
> carefully before modifying input data.

Now, upsert the input data:

```python
from gettsim_personas import upsert_input_data

> upserted_input_data = upsert_input_data(

Reviewer 2: Apologies for the back and forth, but this example looks a lot like we should make it a method of `Persona` and return a copy of the persona. Or am I missing some use case?

```python
from gettsim_personas import get_personas

jan_2025 = "2025-01-01"

personas = get_personas(jan_2025)

> persona = personas.couple_1_child_no_means_tested_transfers

Reviewer 2: Staring at this again, I am not terribly fond of the interface workflow

1. Set up all personas
2. Get one with default values
3. Modify persona

I am not sure whether it would be too large a change, but would the following work?

```py
persona = personas.einkommensteuer_sozialabgaben.couple_1_child(date=jan_2025)
```

We could then easily have users overriding directly there, e.g., using the example from my other comment:

```py
persona = personas.einkommensteuer_sozialabgaben.couple_1_child(
    date=jan_2025,
    n_points=101,
    bruttolohn_m={
        0: {"bottom": 1_500, "top": 10_000},
        1: {"bottom": 1_000, "top": 2_500}
    }
)
```

We would loose being able to see all active personas at one point in time. Instead, you would see all of them and get an `ImplementationError` if requesting a persona for an unsupported date.

Reviewer 2: I don't actually see a big difference. That is, even under that view I don't think it would be terribly helpful to create all too many personas that exist only for a year or two -- how would you ever find them?

I think it depends on customizing the `NotImplementedErrors`. E.g., for `geringverdiener_und_arbeitslose.couple_1_child_in_sgbii_karenzzeit(date="2021-01-01")`, you would get "This persona does not exist before 2023 because there was no grace period (Karenzzeit). Please use  `geringverdiener_und_arbeitslose.couple_1_child` directly.

For `rentner.single_99_prozent_teilrente("2021-01-01")`, you may see: "This persona has not been implemented before 2022 because nobody has had the need yet. Feel free to expand the period for this persona, see [link] for getting started".
